Add metadata to plots (#11765)

AnishereMariam opened the pull request:
@bryevdv, I fixed all the changes and erased every unrelated issue to this PR as requested.

issues: ref #11765

AnishereMariam committed:
Added metadata to filtering.py

AnishereMariam committed:
Added metadata to customjs_expr.py

bryevdv commented on examples/plotting/filtering.py:
```diff
@@ -1,3 +1,15 @@
+''' A map representation of unemployment rate in US using the `US_Sates Dataset`_.
+This example demonstrates using IndexFilter, ColorMapper and HoverTool with
+basic plot elements such as patches. When hovering over the points,
+the state and its umemployment rate is shown.
+.. bokeh-example-metadata::
```
blank like for readability / consistency with existing examples
```suggestion
the state and its umemployment rate is shown.

.. bokeh-example-metadata::
```

bryevdv commented on examples/plotting/customjs_expr.py:
```diff
@@ -1,3 +1,9 @@
+''' This example demonstrates using CustomJS Expression to create a sinusoidal line chart with Bokeh.
+.. bokeh-example-metadata::
+    :apis: bokeh.models.CustomJSExpr, bokeh.model.DataModel
```
blank like for readability / consistency with existing examples
```suggestion
.. bokeh-example-metadata::
    :apis: bokeh.models.CustomJSExpr, bokeh.model.DataModel
```

bryevdv reviewed (changes requested):
One last comments to make things consistent with other existing examples formatting

AnishereMariam committed:
Update metadata `filtering.py`

Co-authored-by: Bryan Van de Ven <bryan@bokeh.org>

AnishereMariam committed:
Update metadata `customjs_expr.py` (#11765)

AnishereMariam committed:
Add metadata `image_url.py` (#11765)

bryevdv commented:
@anisheremariam looks like the linter is still reporting one trailing space issue
> E       AssertionError: Code quality issues:
E         File contains trailing whitespace: examples/basic/scatters/image_url.py, line 7.
E       assert 1 == 0
E        +  where 1 = len(['File contains trailing whitespace: examples/basic/scatters/image_url.py, line 7.'])

FYI you can run all the codebase tests to check these things locally: 

https://docs.bokeh.org/en/latest/docs/dev_guide/testing.html#run-codebase-tests

AnishereMariam committed:
Remove all trailing spaces `image_url.py`

AnishereMariam commented:
@bryevdv, I removed all the trailing spaces and ran all codebase tests locally as requested.

bryevdv commented on examples/basic/scatters/image_url.py:
```diff
@@ -1,3 +1,12 @@
+
+.. bokeh-example-metadata::
+    :apis: bokeh.plotting.figure.image_url
+    :refs: :ref:`ug_basic_scatters_urls`
```
unfortunately this ref doesn't actually exist yet. Either it needs to be removed altogether, or else it must be added to the docs ReST source:

https://raw.githubusercontent.com/bokeh/bokeh/branch-3.2/docs/bokeh/source/docs/user_guide/basic/scatters.rst

It would go right above the section header: 

<img width="379" alt="Screenshot 2023-03-21 at 19 37 54" src="https://user-images.githubusercontent.com/1078448/226787731-5bbb241d-ae2d-43b5-805f-06817f4ac58a.png">

Unfortunately GitHub tries to "helpfully" render ReST files so I to link the raw file and show an image

bryevdv commented on examples/plotting/customjs_expr.py:
```diff
@@ -1,3 +1,10 @@
+''' This example demonstrates using CustomJS Expression to create a sinusoidal line chart with Bokeh.
```
"with Bokeh" is redundant :) 
```suggestion
''' This example demonstrates using CustomJS Expression to create a sinusoidal line chart.
```

bryevdv commented on examples/plotting/customjs_expr.py:
```diff
@@ -1,3 +1,10 @@
+.. bokeh-example-metadata::
+    :apis: bokeh.models.CustomJSExpr, bokeh.model.DataModel
+    :refs: :ref:`ug_interaction_js_callback`
+    :keywords: CustomJS, CustomJSExpr
```
we generally avoid re-iterating the APIs in keywords
```suggestion
    :keywords: js callbacks, expressions
    
```
also a blank line after metadata is typical

bryevdv commented on examples/plotting/filtering.py:
```diff
@@ -1,3 +1,16 @@
+.. bokeh-example-metadata::
+    :sampledata: US_States
+    :apis: bokeh.models.IndexFilter, bokeh.models.HoverTool, bokeh.models.LinearColorMapper
+    :refs: :ref:`ug_basic_areas_patches`, :ref:`ug__basic_data_filtering`
```
typo
```suggestion
    :refs: :ref:`ug_basic_areas_patches`, :ref:`ug_basic_data_filtering`
```

bryevdv commented on examples/plotting/filtering.py:
```diff
@@ -1,3 +1,16 @@
+''' A map representation of unemployment rate in US using the `US_Sates Dataset`_.
```
```suggestion
''' A map representation of unemployment rate in US using the `US_States Dataset`_.
```

bryevdv commented on examples/plotting/filtering.py:
```diff
@@ -1,3 +1,16 @@
+    :refs: :ref:`ug_basic_areas_patches`, :ref:`ug__basic_data_filtering`
+    :keywords: hover tool, indexfilter, filtering, US
+
+.. _US_Sates Dataset: https://docs.bokeh.org/en/latest/docs/reference/sampledata.html#module-bokeh.sampledata.us_states
```
```suggestion
.. _US_States Dataset: https://docs.bokeh.org/en/latest/docs/reference/sampledata.html#module-bokeh.sampledata.us_states
```

bryevdv reviewed (changes requested):
Thank @anisheremariam a few more comments

AnishereMariam committed:
Update metadata `customjs_expr.py`

Co-authored-by: Bryan Van de Ven <bryan@bokeh.org>

AnishereMariam committed:
Add ref link 'scatter.rst'

AnishereMariam committed:
Add ref link to `IndexFilter`

AnishereMariam committed:
Update metadata `filtering.py`

AnishereMariam committed:
Update metadata `filtering.py`

AnishereMariam committed:
Update metadata `customjs_expr.py`

AnishereMariam committed:
Update metadata `customjs_expr.py`

AnishereMariam commented:
@bryevdv, I added the missing references to the doc ReST source,  updated some of the references and fixed the typos in the metadata. Please review for approval.

bryevdv commented on examples/plotting/customjs_expr.py:
```diff
@@ -1,3 +1,11 @@
+.. bokeh-example-metadata::
+    :apis: bokeh.models.CustomJSExpr, bokeh.model.DataModel
+    :refs: :ref:`ug_interaction_js_callbacks_customjs`
+    :keywords: js callbacks, expressions
+'''
```
```suggestion
    :keywords: js callbacks, expressions
    
'''
```

bryevdv commented:
FYI @mattpap it looks like the new test for global ids is failing periodically:
```
FAILED tests/unit/bokeh/util/test_util__serialization.py::test_make_globally_unique_css_safe_id - AssertionError: assert 39 == 36
[421](https://github.com/bokeh/bokeh/actions/runs/4488972948/jobs/7894378460?pr=12979#step:5:422)
 +  where 39 = len('bk-d55c8d77-0b72-48b7-a41b-92bf55d3ecba')
[422](https://github.com/bokeh/bokeh/actions/runs/4488972948/jobs/7894378460?pr=12979#step:5:423)
 +    where 'bk-d55c8d77-0b72-48b7-a41b-92bf55d3ecba' = <function make_globally_unique_css_safe_id at 0x00000231045D3550>()
[423](https://github.com/bokeh/bokeh/actions/runs/4488972948/jobs/7894378460?pr=12979#step:5:424)
 +      where <function make_globally_unique_css_safe_id at 0x00000231045D3550> = bus.make_globally_unique_css_safe_id
```

bryevdv commented:
Thanks @anisheremariam the new references are a great addition!
